// 2d/point/web/Point.js
import * as PIXI            from 'pixi.js'
import React, { Component } from 'react'


import Context              from '2d/stage/all/Context'


class Point extends Component {
  static defaultProps = {
    x: 0,
    y: 0,
  }

  constructor(props) {
    super(props)
  }

  componentDidMount() {
    const { context } = this.props
    const { stage } = context.app

    this.stage = stage

    this.mount()
  }

  componentDidUpdate() {
    this.update()
  }

  componentWillUnmount() {
    this.unmount()
  }

  mount = (payload) => {
    const instance = this.instance = new PIXI.Graphics()

    this.update(this.props)

    instance.interactive = true
    instance.buttonMode = true

    instance.hitArea = new PIXI.Circle(0, 0, 8);

    instance.click = this.onClick
    instance.mouseover = this.onOver
    instance.mouseout = this.onOut

    instance.on('pointerdown', this.onDragStart)
    instance.on('pointerup', this.onDragEnd)
    instance.on('pointerupoutside', this.onDragEnd)
    instance.on('pointermove', this.onDragMove)

    this.stage.addChild(this.instance)
  }

  unmount = (payload) => {
    this.stage.removeChild(this.instance)
  }

  update = (props) => {
    const { x, y } = this.props
    const { color } = this.props
    const instance = this.instance

    instance.position.set(x, y)

    instance.clear()
    
    instance.beginFill(color ? color : 0xffffff)
    instance.lineStyle(2, 0x0000ff, 1)
    instance.drawCircle(0, 0, 8)
    instance.endFill()
  }

  onClick = (payload) => {
    const { context } = this.props

    const { onClick } = this.props
    const {
      data: {
        global: {
          x, y,
        }
      }
    } = payload

    const isDragging = Date.now() - this.timeDown > 300

    onClick && !isDragging && !this.move && onClick({ x, y })
  }

  onOver = () => {
    const { onOver } = this.props

    onOver && onOver()
  }

  onOut = () => {
    const { onOut } = this.props

    onOut && onOut()
  }

  onDragStart = (event) => {
    this.data = event.data
    this.timeDown = Date.now()
    this.alpha = 0.5
    this.dragging = true
    this.move = false
  }

  onDragEnd = () => {
    this.alpha = 1
    this.timeDown = null
    this.dragging = false
    this.data = null
    this.move = false
  }

  onDragMove = () => {
    if (this.dragging) {
      const { x, y } = this.data.getLocalPosition(this.instance.parent)
      
      this.move = true

      const { onDragMove } = this.props

      onDragMove && onDragMove({ x, y })
    }
  }  

  render() {
    return null
  }
}

export default (props) => {
  return (
    <Context.Consumer>
      {(context) => (
        <Point {...props} context={ context } />
      )}
    </Context.Consumer>
  )
}
