// pixi/web/Stage.js
import * as PIXI from 'pixi.js'

import React, { Component } from 'react'

import styled               from 'styled-components'


import Context              from 'pixi/all/Context'
import Provider             from 'pixi/all/Provider'


const who = [`pixi`, `web`, `Stage`]

class Stage extends Component {
  static defaultProps = {
    antialias: true,
    width: 700,
    height: 500,
  }

  constructor() {
    super()

    this.state = {
      isReady: false,
    }
  }

  componentDidMount() {
    const { context } = this.props
    const { antialias, width, height, options } = this.props

    this.app = new PIXI.Application(width, height, {
      ...options,
      
      antialias,
      backgroundColor: 0xffaaaa, // this.app.renderer.backgroundColor
      view: this.canvas,
    })

    const { action } = context

    action.app.set({ app: this.app })

    this.setState({ isReady: true })
  }

  componentDidCatch(error, errorInfo) {
    console.error(...who, `error`, error, `errorInfo`, errorInfo)
  }

  componentWillUnmount() {
    this.renderStage()
    this.app.destroy()
  }  

  renderStage() {
    this.app.renderer.render(this.app.stage)
  }

  onClick = () => {
    const { context } = this.props

    context.debug && console.log(...who, `onClick`)    
  }

  render() {
    return (
      <React.Fragment>
        <canvas
          ref={(canvas) => this.canvas = canvas}
          
          style={{ border: '1px solid black' }}

          onClick={ this.onClick }
        />
        { this.state.isReady && this.props.children }
      </React.Fragment>
    )
  }
}

export default (props) => {
  return (
    <Provider>
      <Context.Consumer>
        {(context) => (
          <Stage {...props} context={ context } />
        )}
      </Context.Consumer>
    </Provider>
  )
}
