// evokeme/space/api/item/remove.js
import _update              from 'immutability-helper'

import { takeEvery }        from 'redux-saga/effects'
import { put }              from 'redux-saga/effects'


_update.extend('$exclude', (excludeList, list) => {
  return list.filter(item => !excludeList.includes(item))
})

export default ({ name, processName = `REMOVE` }) => {
  const processNameDone = `${ processName }_DONE`

  const types = {
    [processName]:      `${ name }/${ processName }`,
    [processNameDone]:  `${ name }/${ processNameDone }`,
  }

  const reducerMap = {
    [types[processNameDone]]: (state, { payload, options }) => {
      const { 
        ns = `default`,
        entity = `default`,
      } = options || {}

      try {
        return _update(state, {
          [ns]: {
            [entity]: {          
              data: {
                $unset: [payload.id],
              },
              order: {
                $exclude: [payload.id],
              },
            },
          },
        })
      } catch (err) {
        return state
      }
    }
  }

  function* processRemove({ payload: { id }, options }) {
    try {
      yield put({ type: types[processNameDone],  payload: { id }, options })
    } catch (error) {
      console.log(error)
    }
  }

  function* saga() {
    yield takeEvery(types[processName], processRemove);
  }

  return {
    reducerMap,
    saga,
    types,
  }
}
