// exteriorAr/page/web/PageHome.js
import React, { Component } from 'react'

import { Button } from 'semantic-ui-react'
import { Container } from 'semantic-ui-react'
import { Input } from 'semantic-ui-react'

import uuid from 'uuid/v4'


import Block from 'layout/web/Block'

import PointContext from 'evokeme/point/all/Context'

import Background from 'pixi/web/Background'
import Line from 'pixi/web/Line'
import Point from 'pixi/web/Point'
import Stage from 'pixi/web/Stage'
import Text from 'pixi/web/Text'
import Texture from 'pixi/web/Texture'

import PointList from 'pixi/web/PointList'
import LineList from 'pixi/web/LineList'
import GroupList from 'pixi/web/GroupList'
import Perspective from 'pixi/web/Perspective'


const who = [`page`, `web`, `PageHome`]

class PageHome extends Component {
  constructor() {
    super()

    this.state = {
      scale: `1`,
    }
  }

  onMountPointList = ({ pointContext }) => {

  }

  onMountGroupList = ({ context }) => {
    console.log(...who, `onMountLineList`, context)

    context.point.create({ id: `point1`, x: 50, y: 100, })
    context.point.create({ id: `point2`, x: 50, y: 200, })
    context.point.create({ id: `point3`, x: 150, y: 200, })
    context.point.create({ id: `point4`, x: 150, y: 100, })

    context.point.create({ id: `p1`, x: 300, y: 400, })
    context.point.create({ id: `p2`, x: 400, y: 400, })
    context.point.create({ id: `p3`, x: 400, y: 500, })
    context.point.create({ id: `p4`, x: 300, y: 500, })

    context.group.create({ id: `group1`, name: `line1`, })
    
    context.group.create({
      id: `perspective`,
      name: `perspective`,
      
      isHide: true,
    })

    context.groupToPoint.create({ lineId: `group1`, pointId: `point1`, })
    context.groupToPoint.create({ lineId: `group1`, pointId: `point2`, })
    context.groupToPoint.create({ lineId: `group1`, pointId: `point3`, })
    context.groupToPoint.create({ lineId: `group1`, pointId: `point4`, })

    context.groupToPoint.create({ lineId: `perspective`, pointId: `p1`, })
    context.groupToPoint.create({ lineId: `perspective`, pointId: `p2`, })
    context.groupToPoint.create({ lineId: `perspective`, pointId: `p3`, })
    context.groupToPoint.create({ lineId: `perspective`, pointId: `p4`, })
  }


  onChange = ({ target: { name, value }}) => this.setState({ [name]: value })

  render() {
    const { scale } = this.state

    return (
      <Block middle center style={{ height: '100vh' }}>
        <Stage>
          <Background
            imageLink={`https://i.imgur.com/jwew23c.jpg`}
            onClick={ this.onClick }
          />

          <GroupList
            scale={ Number(scale) || 0 }

            onMount={ this.onMountGroupList }
          />

          <Perspective />

          <PointList />
        </Stage>
        <Block left style={{ marginBottom: 10 }}>
          <Button content='Масштаб' primary />
          <Input
            name='scale'
            placeholder=''
            value={ scale || `` }

            onChange={ this.onChange }
          />
        </Block>
      </Block>
    )
  }
}

export default PageHome
