// core/mutation/api/start.js
import _update              from 'immutability-helper'

import { takeEvery }        from 'redux-saga/effects'
import { put }              from 'redux-saga/effects'


import spaceTypes           from 'core/space/api/types'
import mutationTypes        from 'core/mutation/api/types'


export default ({ name, processName = `START` }) => {
  const processNameDone = `${ processName }_DONE`

  const types = {
    [processName]:      `${ name }/${ processName }`,
    [processNameDone]:  `${ name }/${ processNameDone }`,
  }

  const reducerMap = {
    [types[processNameDone]]: (state, { payload }) => {
      const { name } = payload

      return _update(state, {
        $merge: {
          [name]: {
            isLoading: true,
          },
        }
      })
    }
  }

  function* processStart({ payload, options }) {
    try {
      const { mutation } = payload
      const { props } = payload
      const { isUpdate = true } = options || {}

      yield put({
        type: types[processNameDone],
        payload
      })

      const { data, order } = yield mutation(props)

      yield put({
        type: isUpdate ? spaceTypes.UPDATE : spaceTypes.DELETE,
        payload: {
          data,
          order,
        },
        options,
      })

      yield put({
        type: mutationTypes.FINISH,
        payload,
      })
    } catch (error) {
      console.log(error)
    }
  }

  function* saga() {
    yield takeEvery(types[processName], processStart)
  }

  return {
    reducerMap,
    saga,
    types,
  }
}
