// evokeme/redux/api/create.js
import _update              from 'immutability-helper'

import { takeEvery }        from 'redux-saga/effects'
import { put }              from 'redux-saga/effects'

import uuid                 from 'uuid/v4'


export default ({ name, processName = `CREATE` }) => {
  const processNameDone = `${ processName }_DONE`

  const types = {
    [processName]:      `${ name }/${ processName }`,
    [processNameDone]:  `${ name }/${ processNameDone }`,
  }

  const reducerMap = {
    [types[processNameDone]]: (state, { payload, options }) => {
      const { 
        ns = `default`,
        entity = `default`,
      } = options

      return _update(state, {
        [ns]: {
          [entity]: {
            data: {
              $merge: { [payload.id]: payload },
            },
            order: {
              $push: [payload.id],
            },
          },
        },
      })
    }
  }

  function* processCreate({ payload, options }) {
    try {
      const id = payload.id || uuid()
      yield put({ type: types[processNameDone], payload: { ...payload, id }})
    } catch (error) {
      console.log(error)
    }
  }

  function* saga() {
    yield takeEvery(types[processName], processCreate)
  }

  return {
    reducerMap,
    saga,
    types,
  }
}
