// core/redux/api/create.js
import _update              from 'immutability-helper'

import { takeEvery }        from 'redux-saga/effects'
import { put }              from 'redux-saga/effects'
import { select }           from 'redux-saga/effects'

import uuid                 from 'uuid/v4'


const reduceEntity = (
  state = {
    data: {},
    order: [],
  },
    { payload }
) => {
  const { data } = state

  return _update(state, {
    data: {
      $merge: payload.data,
    },
    order: {
      $push: payload.order.filter((id) => !data[id]),
    },
  })
}

const reduceSpace = (state = {}, { payload, options }) => {
  const { entity = `default` } = options

  return {
    ...state,
    [entity]: reduceEntity(state[entity], { payload, options })
  }
}

export default ({ name, processName = `CREATE` }) => {
  const processNameDone = `${ processName }_DONE`

  const types = {
    [processName]:      `${ name }/${ processName }`,
    [processNameDone]:  `${ name }/${ processNameDone }`,
  }

  const reducerMap = {
    [types[processNameDone]]: (state, { payload, options = {} }) => {
      const { ns = `default` } = options

      return {
        ...state,
        [ns]: reduceSpace(state[ns], { payload, options })
      }
    }
  }

  function* processCreate({ payload, options }) {
    try {
      const { id = uuid() } = payload

      yield put({
        type: types[processNameDone],
        payload: { ...payload, id },
        options,
      })
    } catch (error) {
      console.log(error)
    }
  }

  function* saga() {
    yield takeEvery(types[processName], processCreate)
  }

  return {
    reducerMap,
    saga,
    types,
  }
}
